src/Prototype/Types/Secret.hs

```diff
@@ -0,0 +1,102 @@
+{-# LANGUAGE
+    TypeFamilies
+  , TypeOperators
+  , UndecidableInstances
+#-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE
+    TypeSynonymInstances
+  , FlexibleInstances
+  , FlexibleContexts
+#-}
+{-# LANGUAGE
+    DerivingStrategies
+  , DeriveAnyClass
+  , DerivingVia 
+  , DeriveGeneric 
+  , GeneralizedNewtypeDeriving
+  , StandaloneDeriving
+#-}
+
+module Prototype.Types.Secret
+  ( Secret(..)
+  , SecretExp(..)
+  , exposeSecret
+  ) where
+
+import           Control.Lens
+import           Data.Aeson
+import qualified GHC.Show                      as Show
+import qualified GHC.TypeLits                  as TL
+import           Servant.API                    ( FromHttpApiData )
+
+-- | Exposure level of the secret.
+data SecretExp = ToJSONExp
+  deriving (Eq, Show)
+
+-- | Constraint: result in type errors when not met, empty otherwise.
+type family HasExp (exp :: SecretExp) (exps :: [SecretExp]) :: Constraint where
```

Comment:
- You do like recursion at the type level :)
- I'll have to play with it, this is still a bit magical to me.

Reply:
Yes; this kind of type-encoded proofs are awesome; and they help a lot with peace of mind. 
The syntax, at least in Haskell, takes some getting used to (backticks); but other than that; type-level proofs are just functions at the type-level. 
Crazy stuff can be achieved with this :-)